### PR TITLE
Bidding states

### PR DESCRIPTION
**Pull request by ashfurrow**
Re: #1684, #1695.

Edit: I've added for to support jump bids to always show our bid at the top in https://github.com/artsy/eigen/pull/1699/commits/6d850d9ea7389832011b354a5686e04b6e558195 

I think we get in weird states because we're trying to do too much. Prediction shows you that you're the highest bidder, even before your bid is accepted by the operator:

screen shot 2016-06-11 at 1 50 39 pm
screen shot 2016-06-11 at 1 50 34 pm

To finish this PR, I'm going to replicate that behaviour.

### CODE REVIEW

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
          3 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">No CHANGELOG changes made</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">Developer Specific file shouldn&#39;t be changed</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">There were <a href="https://eigen-ci.s3.amazonaws.com/2928/index.html">snapshot errors</a></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
  </tbody>
</table>

Time | Class | Function |
| --- | ----- | ----- |237.4ms | [AuctionViewController.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Auction/AuctionViewController.swift#L183) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
189.6ms | [SwiftyJSON.swift](/artsy/eigen/tree/ashfurrow-fixes/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
172.7ms | [WebSocket.swift](/artsy/eigen/tree/ashfurrow-fixes/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
78.3ms | [Observable.swift](/artsy/eigen/tree/ashfurrow-fixes/Pods/Interstellar/Sources/Observable.swift#L71) | private final func nextTokenHash() -> Int
71.7ms | [AuctionLotMetadataStackScrollView.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Live_Auctions/Views/AuctionLotMetadataStackScrollView.swift#L18) | required init(viewModel: LiveAuctionLotViewModelType, sideMargin: String)
71.1ms | [LiveAuctionsAdminViewController.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Live_Auctions/LiveAuctionsAdminViewController.swift#L22) | @objc override func viewDidLoad()
65.5ms | [LiveAuctionLotViewController.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift#L49) | @objc override func viewDidLoad()
63.6ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L25) | func ar_presentBlurredOverlayWithTitle(title: String, subtitle: String, buttonState: BlurredStatusOverlayViewCloseButtonState = default)
61.7ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L31) | (closure)
59.5ms | [RefinementOptionsViewController+Private.swift](/artsy/eigen/tree/ashfurrow-fixes/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift#L63) | func stackView() -> ORStackView


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

File: Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift [hunk 1]
```diff
@@ -95,7 +95,7 @@ class LiveAuctionLotViewModel: NSObject, LiveAuctionLotViewModelType {
     let newEventsSignal = Observable<[LiveAuctionEventViewModel]>()
 
     var sellingToBidderID: String? = nil
-    var winningBidderID: String? = nil
+    var winningBidEventId: String? = nil
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift]: We were calculating the winning bid event before, but causality provides it for us, so we should use it.

**Comment by ashfurrow** [comment]
This PR brings eigen more in line with how prediction handles bid history and bid button states. When a user submits a bid, we used to have a "bid pending" state while we wait for the operator to accept the bid. Instead, we now assume they're the highest bidder right away, until they get outbid.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift [hunk 2]
```diff
@@ -163,7 +163,7 @@ class LiveAuctionPlaceMaxBidViewController: UIViewController {
         bidViewModel.currentBid = max(bidViewModel.currentBid, bidViewModel.currentLotValue)
         updateBiddingControls(bidViewModel.currentBid)
 
-        shouldShowBiddingOverlay(false, maxBidder: bidViewModel.lotViewModel.userIsHighestBidder)
+        shouldShowBiddingOverlay(false, maxBidder: bidViewModel.lotViewModel.userIsBeingSoldTo)
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionPlaceMaxBidViewController.swift]: The `userIsHighestBidder` was changed to no longer be calculated by us, but defined by causality. Our other logic depended on having the events added _then_ all derived state updated, but we now update user-specific derived state first, so this line had to change.

Ultimately I think having a better atomic change operation would avoid having to perform updates in a careful order. Definitely an area to improve this codebase in.

File: Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionEventViewModel.swift [hunk 3]
```diff
@@ -96,10 +104,6 @@ class LiveAuctionEventViewModel: NSObject, LiveAuctionEventViewModelType {
                 color = .blackColor()
             }
 
-            // Override color when an unconfirmed Artsy Bidder
-            let isUnConfirmedArtsyBidder = isArtsyBidder && !event.confirmed
-            color = isUnConfirmedArtsyBidder ? .artsyGrayMedium() : color
-
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionEventViewModel.swift]: This was removed to bring eigen in line with how prediction works, confirmed with @devangt this behaviour is the goal.

**Comment by ashfurrow** [comment]
Snapshot changes I think are caused by the atomic change issue I commented on earlier; it's only an off-by-one difference on a non-critical part of the UI, and I don't know which is correct from the stored json (before or after) so I just swapped 'em.

**Comment by orta** [comment]
Cool, this looks good to me

**Pull request merged**